File: tech-shop-api/src/main/java/md/ceiti/techshopapi/entity/product/Product.java
package md.ceiti.techshopapi.entity.product;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import md.ceiti.techshopapi.util.ConstraintViolationMessage;
import org.hibernate.annotations.Check;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Entity
@Getter
@Setter
@ToString
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true, columnDefinition = "varchar(255)")
    @NotBlank(message = ConstraintViolationMessage.NOT_BLANK)
    @Size(
            min = 10,
            max = 255,
            message = ConstraintViolationMessage.SIZE
    )
    private String name;

    @NotNull(message = ConstraintViolationMessage.NOT_NULL)
    @ManyToOne
    private Category category;

    @Column(nullable = false)
    @NotNull(message = ConstraintViolationMessage.NOT_NULL)
    private Double price;

    @Column(columnDefinition = "varchar(8)")
    @Enumerated(value = EnumType.STRING)
    private DiscountType discountType;

    @Check(
            constraints = "(discount_type IS NULL AND discount IS NULL) OR " +
                          "(discount_type = 'AMOUNT' AND discount > 0 AND discount < price) OR " +
                          "(discount_type = 'PERCENT' AND discount > 0 AND discount < 100)"
    )
    private Double discount;

    @Column(nullable = false)
    @NotNull(message = ConstraintViolationMessage.NOT_NULL)
    private Integer available;

    /**
     * Must contain producer, series, model
     * */

    @ElementCollection
    private Map<String, String> general;

    /**
     * Must contain at least one
     * */

    @ElementCollection
    @NotEmpty(message = ConstraintViolationMessage.NOT_EMPTY)
    private List<byte[]> images;

    /**
     * Must contain diagonal, resolution
     * */

    @ElementCollection
    private Map<String, String> display;

    /**
     * Must contain processor
     * */

    @ElementCollection
    private Map<String, String> processor;

    /**
     * Must contain RAM, ROM
     * */

    @ElementCollection
    private Map<String, String> memory;

    /**
     * Must contain GPU, type
     * */

    @ElementCollection
    private Map<String, String> gpu;

    /**
     * Must contain ports, connectivity
     * */

    @ElementCollection
    private Map<String, String> functionality;

    /**
     * Must contain wifi, bluetooth
     * */

    @ElementCollection
    private Map<String, String> connectivity;
}

File: tech-shop-app/src/app/model/product.ts
export class ProductCardResponse {
  id: number;
  name: string;
  price: number;
  discountType: string;
  discount: number;
  specs: string;
  image: any;


  constructor(id: number, name: string, price: number, discountType: string, discount: number, specs: string, image: any) {
    this.id = id;
    this.name = name;
    this.price = price;
    this.discountType = discountType;
    this.discount = discount;
    this.specs = specs;
    this.image = image;
  }
}

export class Product {
  id: number;
  name: string;
  category: Category;
  price: number;
  discountType: string;
  discount: number;
  general: Map<string, string>;
  display: Map<string, string>;
  processor: Map<string, string>;
  memory: Map<string, string>;
  gpu: Map<string, string>;
  functionality: Map<string, string>;
  get categoryName(): string {
    return this.category.name + (this.category.parent !== null ? (' ' + this.category.parent.name) : '');
  }

  get fullDiscount(): string {
    return this.discount + (this.discountType === 'PERCENT' ? '%' : '');
  }
}

export class Category {
  id: number;
  parent: Category;
  name: string;
}

export class Laptop {
  id: number;
  name: string;
  category: Category;
  price: number;
  discountType: string;
  discount: number;
  general: Map<string, string>;
  display: Map<string, string>;
  processor: Map<string, string>;
  memory: Map<string, string>;
  gpu: Map<string, string>;
  functionality: Map<string, string>;
}
